Applications/Utils/MeshEdit/AddTopLayer.cpp

```diff
@@ -0,0 +1,175 @@
+#include "MeshLib/Elements/Element.h"
+#include "MeshLib/Elements/Prism.h"
+#include "MeshLib/Elements/Hex.h"
+#include "MeshLib/Elements/Quad.h"
+
+// MeshGeoToolsLib
+#include "MeshGeoToolsLib/MeshNodeSearcher.h"
+
+int main (int argc, char* argv[])
+{
+	LOGOG_INITIALIZE();
+	logog::Cout* logog_cout (new logog::Cout);
+	BaseLib::LogogSimpleFormatter *custom_format (new BaseLib::LogogSimpleFormatter);
+	logog_cout->SetFormatter(*custom_format);
+
+	TCLAP::CmdLine cmd(
+		"Adds a top layer to an existing 3d mesh",
+		' ',
+		"0.1");
+
+	TCLAP::ValueArg<std::string> mesh_arg("m", "mesh-file",
+		"the name of the file containing the mesh", true,
+		"", "file name");
+	cmd.add(mesh_arg);
+
+	TCLAP::ValueArg<std::string> mesh_out_arg("o", "output-mesh-file",
+		"the name of the file the mesh should be written to (vtu format)", true,
+		"", "file name");
+	cmd.add(mesh_out_arg);
+
+	TCLAP::ValueArg<double> layer_thickness_arg("t", "layer-tickness",
+		"the thickness of the new layer", false, 10, "floating point value");
+	cmd.add(layer_thickness_arg);
+
+	cmd.parse(argc, argv);
+
+	INFO("Reading mesh \"%s\" ... ", mesh_arg.getValue().c_str());
+	MeshLib::Mesh * subsfc_mesh(FileIO::readMeshFromFile(mesh_arg.getValue()));
+	INFO("done.");
+
+	INFO("Extracting top surface of mesh \"%s\" ... ",
+		mesh_arg.getValue().c_str());
+	const MathLib::Vector3 dir(0,0,-1);
+	double const angle(90);
+	std::unique_ptr<MeshLib::Mesh> sfc_mesh(
+		MeshLib::MeshSurfaceExtraction::getMeshSurface(
+			*subsfc_mesh, dir, angle, true
+		)
+	);
+	INFO("done.");
+
+	// *** add new top surface nodes
+	std::vector<MeshLib::Node*> & subsfc_nodes(
+		const_cast<std::vector<MeshLib::Node*> &>(subsfc_mesh->getNodes())
+	);
+	std::size_t const n_subsfc_nodes(subsfc_nodes.size());
+
+	std::vector<MeshLib::Node*> const& sfc_nodes(sfc_mesh->getNodes());
+	std::size_t const n_sfc_nodes(sfc_nodes.size());
+
+	double const layer_thickness(layer_thickness_arg.getValue());
+	// *** copy sfc nodes to subsfc mesh node
+	std::map<std::size_t, std::size_t> subsfc_sfc_id_map;
+	for (std::size_t k(0); k<n_sfc_nodes; ++k) {
+		subsfc_nodes.push_back(new MeshLib::Node(*sfc_nodes[k]));
+		std::size_t const subsfc_id(sfc_nodes[k]->getID());
+		std::size_t const sfc_id(k+n_subsfc_nodes);
+		subsfc_sfc_id_map.insert(std::make_pair(subsfc_id, sfc_id));
+		(*(subsfc_nodes.back()))[2] += layer_thickness;
+	}
+	subsfc_mesh->resetNodeIDs();
+
+	// *** insert new top layer elements into subsfc_mesh
+	std::vector<MeshLib::Element*> & subsfc_elements(
+		const_cast<std::vector<MeshLib::Element*> &>(subsfc_mesh->getElements())
+	);
+	std::size_t orig_size(subsfc_elements.size());
+	std::vector<MeshLib::Element*> const& sfc_elements(sfc_mesh->getElements());
+	std::size_t const n_sfc_elements(sfc_elements.size());
+	for (std::size_t k(0); k<n_sfc_elements; ++k) {
+		MeshLib::Element const*const sfc_elem(sfc_elements[k]);
+		if (sfc_elem->getGeomType() == MeshElemType::TRIANGLE) {
+			// add a prism
```

Comment:
I'd extract the innermost if into own function.

Is an element of POINT type not implemented or is it just missing here?